### PR TITLE
Sync all Collection versions

### PR DESCRIPTION
**Pull request by fao89**
We were not considering pagination, we just grabbed the first 10 versions

https://pulp.plan.io/issues/7740
closes #7740

### CODE REVIEW

File: pulp_ansible/tests/functional/api/collection/v2/test_sync.py [hunk 1]
```diff
@@ -253,6 +253,31 @@ def test_mirror_sync_with_invalid_requirements(self):
 
         self.assertEqual(collection_remote.status_code, 400, collection_remote)
 
+    def test_sync_with_aws_requirements(self):
```
Comment (fao89) [inline pulp_ansible/tests/functional/api/collection/v2/test_sync.py:256]: Thank you @daviddavis 
Having an issue with a functional test was very helpful

File: CHANGES/7740.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Sync all Collection versions
```
Comment (daviddavis) [inline CHANGES/7740.bugfix:1]: A better changelog message here would be great.

File: pulp_ansible/app/tasks/collections.py [hunk 3]
```diff
@@ -354,26 +353,38 @@ def _build_url(path_or_url):
...
 
-        progress_data = dict(message="Parsing Galaxy Collections API", code="parsing.collections")
-        with ProgressReport(**progress_data) as progress_bar:
-            api_version = get_api_version(remote.url)
-            url = _get_url(page_count, api_version)
+        async def _loop_through_pages(original_url, not_done):
+            """
+            Loop through API pagination.
+            """
+            page_count = 1
```
Comment (daviddavis) [inline pulp_ansible/app/tasks/collections.py:360]:
```suggestion
```

File: pulp_ansible/app/tasks/collections.py [hunk 4]
```diff
@@ -354,26 +353,38 @@ def _build_url(path_or_url):
...
+            api_version = get_api_version(original_url)
+            url = _get_url(page_count, api_version, original_url)
```
Comment (daviddavis) [inline pulp_ansible/app/tasks/collections.py:362]:
```suggestion
            url = _get_url(1, api_version, original_url)
```

File: CHANGES/7740.bugfix (hunk 2 above)
Comment (fao89) [reply CHANGES/7740.bugfix:1]: How does it sound:
```
Enable sync for more than 10 collection versions
```

File: CHANGES/7740.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Sync more than 10 versions per collection
```
Comment (daviddavis) [inline CHANGES/7740.bugfix:1]: How about "Fixed bug where only 10 collections were being synced in some cases"

Comment (fao89) [reply CHANGES/7740.bugfix:1]: much better!

**Comment by daviddavis** [comment]
Code looks good. I tested it out and it works. One strange thing though: the total counts in the progress report seem to be a bit off:

```
      {                                                                                                                                                                             
        "message": "Parsing Collection Metadata",                                                                                                                                   
        "code": "parsing.metadata",                                                                                                                                                 
        "state": "completed",                                                                                                                                                       
        "total": null,                                                                                                                                                              
        "done": 67,
        "suffix": null
      },
      {
        "message": "Parsing Galaxy Collections API",
        "code": "parsing.collections",
        "state": "completed",
        "total": 100,
        "done": 67,
        "suffix": null
      },
      {
        "message": "Downloading Artifacts",
        "code": "downloading.artifacts",
        "state": "completed",
        "total": null,
        "done": 67,
        "suffix": null
      },
      {
        "message": "Downloading Artifacts",
        "code": "downloading.artifacts",
        "state": "completed",
        "total": null,
        "done": 67,
        "suffix": null
      },
            {
        "message": "Associating Content",
        "code": "associating.content",
        "state": "completed",
        "total": null,
        "done": 67,
        "suffix": null
      },
      {
        "message": "Un-Associating Content",
        "code": "unassociating.content",
        "state": "completed",
        "total": null,
        "done": 0,
        "suffix": null
      }
```

Looking at your code though, I think the problem exists prior to this change.

File: pulp_ansible/app/tasks/collections.py [hunk 6]
```diff
@@ -284,7 +284,8 @@ async def run(self):
         """
         Build and emit `DeclarativeContent` from the ansible metadata.
         """
-        with ProgressReport(message="Parsing Collection Metadata", code="parsing.metadata") as pb:
+        msg = "Parsing CollectionVersion Metadata"
+        with ProgressReport(message=msg, code="parsing.metadata") as pb:
```
Comment (fao89) [inline pulp_ansible/app/tasks/collections.py:288]:
```
     "progress_reports": [
                {
                    "message": "Parsing Galaxy Collections API",
                    "code": "parsing.collections",
                    "state": "completed",
                    "total": 1,
                    "done": 1,
                    "suffix": null
                },
                {
                    "message": "Parsing Collection Metadata",
                    "code": "parsing.metadata",
                    "state": "completed",
                    "total": null,
                    "done": 67,
                    "suffix": null
                },
                {
                    "message": "Downloading Artifacts",
                    "code": "downloading.artifacts",
                    "state": "completed",
                    "total": null,
                    "done": 0,
                    "suffix": null
                },
                {
                    "message": "Downloading Artifacts",
                    "code": "downloading.artifacts",
                    "state": "completed",
                    "total": null,
                    "done": 67,
                    "suffix": null
                },
                {
                    "message": "Associating Content",
                    "code": "associating.content",
                    "state": "completed",
                    "total": null,
                    "done": 67,
                    "suffix": null
                },
                {
                    "message": "Un-Associating Content",
                    "code": "unassociating.content",
                    "state": "completed",
                    "total": null,
                    "done": 0,
                    "suffix": null
                }
```

Comment (fao89) [inline pulp_ansible/app/tasks/collections.py:288]:
```
                {
                    "message": "Parsing Galaxy Collections API",
                    "code": "parsing.collections",
                    "state": "completed",
                    "total": 1,
                    "done": 1,
                    "suffix": null
                },
                {
                    "message": "Parsing CollectionVersion Metadata",
                    "code": "parsing.metadata",
                    "state": "completed",
                    "total": null,
                    "done": 67,
                    "suffix": null
                },
                {
                    "message": "Downloading Artifacts",
                    "code": "downloading.artifacts",
                    "state": "completed",
                    "total": null,
                    "done": 0,
                    "suffix": null
                },
                {
                    "message": "Downloading Artifacts",
                    "code": "downloading.artifacts",
                    "state": "completed",
                    "total": null,
                    "done": 67,
                    "suffix": null
                },
                {
                    "message": "Associating Content",
                    "code": "associating.content",
                    "state": "completed",
                    "total": null,
                    "done": 67,
                    "suffix": null
                },
                {
                    "message": "Un-Associating Content",
                    "code": "unassociating.content",
                    "state": "completed",
                    "total": null,
                    "done": 0,
                    "suffix": null
                }
```

**Review by daviddavis — APPROVED** [review]

**Pull request merged**